cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,337 @@
+import functools
+from typing import cast, Any, Dict, List, Sequence, Tuple
+
+from sympy.logic.boolalg import And, Not, Or, Xor
+from sympy.core.expr import Expr
+from sympy.core.symbol import Symbol
+import sympy.parsing.sympy_parser as sympy_parser
+
+import cirq
+from cirq import value
+from cirq.ops import raw_types
+
+
+def _Hamiltonian_O():
+    return 0.0 * cirq.PauliString({})
+
+
+def _Hamiltonian_I():
+    return cirq.PauliString({})
+
+
+def _Hamiltonian_Z(qubit):
+    return cirq.PauliString({qubit: cirq.Z})
+
+
+def _build_hamiltonian_from_boolean(
+    boolean_expr: Expr, qubit_map: Dict[str, 'cirq.Qid']
+) -> 'cirq.PauliString':
+    """Builds the Hamiltonian representation of Boolean expression as per [1]:
+
+    It is essentially a polynomial of Pauli Zs on different qubits. For example, this object could
+    represent the polynomial 0.5*I - 0.5*Z_1*Z_2 and it would be stored inside this object as:
+    self._hamiltonians = {(): 0.5, (1, 2): 0.5}.
+
+    While this object can represent any polynomial of Pauli Zs, in this file, it will be used to
+    represent a Boolean operation which has a unique representation as a polynomial. This object
+    only handle basic operation on the polynomials (e.g. multiplication). The construction of a
+    polynomial from a Boolean is performed by _build_hamiltonian_from_boolean().
+
+    References:
```

Comment:
Let's put these references on the top of the module, that way it shows the "global scope" of the papers.

Reply:
Done.